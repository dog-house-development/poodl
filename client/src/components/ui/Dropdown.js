import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from './Button';
import _ from 'lodash';
import onClickOutside from 'react-onclickoutside';
import classnames from 'classnames';

const propTypes = {
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    kind: PropTypes.oneOf(['primary', 'secondary', 'tertiary']),
    buttonContent: PropTypes.string,
    dropdownContent: PropTypes.array,
    align: PropTypes.oneOf(['left', 'center', 'right']),
    arrow: PropTypes.bool,
    caret: PropTypes.bool,
    icon: PropTypes.string
};

const defaultProps = {
    size: 'medium',
    kind: 'primary',
    align: 'left',
    arrow: false
};

export class Dropdown extends Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false
        };
        this.onDropdownClick = this.onDropdownClick.bind(this);
    }

    handleKeyDown = e => {
        // escape key press
        if (e.keyCode === 27) {
            this.setState({ open: false });
        }
    };

    componentDidMount() {
        document.addEventListener('keydown', this.handleKeyDown);
    }

    componentWillUnmount() {
        document.removeEventListener('keydown', this.handleKeyDown);
    }

    onDropdownClick(e) {
        e.preventDefault();
        this.setState({ open: !this.state.open });
    }

    onDropdownRowClick(onRowClick, evt) {
        evt.preventDefault();
        this.setState({ open: !this.state.open });
        onRowClick(evt);
    }

    handleClickOutside = evt => {
        // should not close dropdown if modal is open
        if (this.state.open && !_.find(evt.path, { id: 'modal' })) {
            this.setState({ open: !this.state.open });
        }
    };

    getDropdownContentMarkup() {
        let arrowMarkup;
        if (this.props.arrow) {
            arrowMarkup = (
                <>
                    <div className="arrow arrow-behind" />
                    <div className="arrow" />
                </>
            );
        }
        if (this.state.open) {
            return (
                <div
                    className={classnames('dropdown-content', `dropdown-align-${this.props.align}`, {
                        'arrow-dropdown': this.props.arrow
                    })}>
                    {arrowMarkup}
                    {this.props.children}
                    {_.map(this.props.dropdownContent, row => {
                        if (_.get(row, 'type') === 'divider') {
                            return <hr key={_.uniqueId('divider-')} />;
                        }
                        return (
                            <button
                                key={_.uniqueId('dropdown-row-')}
                                className={`dropdown-content-row ${this.props.size}`}
                                onClick={evt => this.onDropdownRowClick(row.onClick, evt)}>
                                {row.content}
                            </button>
                        );
                    })}
                </div>
            );
        }
    }

    getCaretMarkup() {
        if (this.props.caret) {
            return (
                <i className="material-icons dropdown-caret">{this.state.open ? 'arrow_drop_up' : 'arrow_drop_down'}</i>
            );
        }
    }

    render() {
        return (
            <div className="dropdown-wrapper">
                <Button
                    size={this.props.size}
                    kind={this.props.kind}
                    onClick={this.onDropdownClick}
                    icon={this.props.icon}>
                    {this.props.buttonContent}
                    {this.getCaretMarkup()}
                </Button>
                {this.getDropdownContentMarkup()}
            </div>
        );
    }
}

Dropdown.propTypes = propTypes;
Dropdown.defaultProps = defaultProps;
export default onClickOutside(Dropdown);
