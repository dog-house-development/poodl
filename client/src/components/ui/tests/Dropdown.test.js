import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { Dropdown } from '../Dropdown';

configure({ adapter: new Adapter() });

describe('Dropdown tests', () => {
    let wrapper, instance;
    const setInstanceAndWrapper = (_props = {}) => {
        const props = _.assign(
            {},
            {
                size: 'medium',
                kind: 'primary',
                buttonContent: 'Dropdown time',
                dropdownContent: [
                    { content: 'party', onClick: () => {} },
                    { type: 'divider' },
                    { content: 'woohoo', onClick: () => {} }
                ]
            },
            _props
        );
        wrapper = shallow(<Dropdown {..._.assign({}, props)} />);
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });
    it('should set state to the open if it is closed', () => {
        instance.setState({ open: false });
        spyOn(instance, 'onDropdownClick');
        const e = {
            preventDefault: () => {}
        };
        instance.onDropdownClick(e);
    });
    describe('onDropdownClick', () => {
        it('should set state to the closed if it is open', () => {
            instance.setState({ open: true });
            spyOn(instance, 'onDropdownClick');
            const e = {
                preventDefault: () => {}
            };
            instance.onDropdownClick(e);
        });
    });

    describe('onDropdownRowClick', () => {
        it('should set state to the open if it is closed', () => {
            instance.setState({ open: false });
            spyOn(instance, 'onDropdownRowClick');
            const e = {
                preventDefault: () => {}
            };
            instance.onDropdownRowClick(e);
        });

        it('should set state to the closed if it is open', () => {
            instance.setState({ open: true });
            spyOn(instance, 'onDropdownRowClick');
            const e = {
                preventDefault: () => {}
            };
            instance.onDropdownRowClick(e);
        });
    });

    describe('handleClickOutside', () => {
        it('should set state to the open if it is closed', () => {
            instance.setState({ open: false });
            spyOn(instance, 'handleClickOutside');
            const e = {
                preventDefault: () => {}
            };
            instance.handleClickOutside(e);
        });

        it('should set state to the closed if it is open', () => {
            instance.setState({ open: true });
            spyOn(instance, 'handleClickOutside');
            const e = {
                preventDefault: () => {}
            };
            instance.handleClickOutside(e);
        });
    });

    describe('getDropdownContentMarkup', () => {
        it('should not return markup because open is false', () => {
            instance.setState({ open: false });
            expect(instance.getDropdownContentMarkup()).toEqual(undefined);
        });
    });

    describe('render', () => {
        it('should render closed dropdown correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should render open dropdown correctly', () => {
            instance.setState({ open: true });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render open arrow dropdown correctly', () => {
            setInstanceAndWrapper({ arrow: true });
            instance.setState({ open: true });
            expect(wrapper).toMatchSnapshot();
        });
    });
});
