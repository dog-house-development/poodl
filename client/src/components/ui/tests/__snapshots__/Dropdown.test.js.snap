// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown tests render should render closed dropdown correctly 1`] = `
<div
  className="dropdown-wrapper"
>
  <Button
    kind="primary"
    onClick={[Function]}
    size="medium"
    type="button"
  >
    Dropdown time
  </Button>
</div>
`;

exports[`Dropdown tests render should render open arrow dropdown correctly 1`] = `
<div
  className="dropdown-wrapper"
>
  <Button
    kind="primary"
    onClick={[Function]}
    size="medium"
    type="button"
  >
    Dropdown time
  </Button>
  <div
    className="dropdown-content dropdown-align-left arrow-dropdown"
  >
    <div
      className="arrow arrow-behind"
    />
    <div
      className="arrow"
    />
    <button
      className="dropdown-content-row medium"
      key="dropdown-row-13"
      onClick={[Function]}
    >
      party
    </button>
    <hr
      key="divider-14"
    />
    <button
      className="dropdown-content-row medium"
      key="dropdown-row-15"
      onClick={[Function]}
    >
      woohoo
    </button>
  </div>
</div>
`;

exports[`Dropdown tests render should render open dropdown correctly 1`] = `
<div
  className="dropdown-wrapper"
>
  <Button
    kind="primary"
    onClick={[Function]}
    size="medium"
    type="button"
  >
    Dropdown time
  </Button>
  <div
    className="dropdown-content dropdown-align-left"
  >
    <button
      className="dropdown-content-row medium"
      key="dropdown-row-10"
      onClick={[Function]}
    >
      party
    </button>
    <hr
      key="divider-11"
    />
    <button
      className="dropdown-content-row medium"
      key="dropdown-row-12"
      onClick={[Function]}
    >
      woohoo
    </button>
  </div>
</div>
`;
